### PR TITLE
Add: Save time theme.json escaping

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
This PR adds save time escaping of the user theme.json.

It tries to follow what was done for HTML post content escaping in src/wp-includes/kses.php.
Like KSES we rely on the content_save_pre filter to escape the content of the post.
Here comes the first challenge, code executing inside of that filter does not have access to the post type being escaped. The filter just receives the content and should output escaped content. So we need some way to given a post content string, identify if it is a user theme.json content.
The way we are doing this is by including a flag isGlobalStylesUserThemeJSON in the user theme.json, which marks if the content of the post is a user theme.json.

Our code totally ignores the user JSON if the flag is not present as in that case the content can not be considered safe.
To filter what CSS is possible we rely on safecss_filter_attr. If a rule is removed by safecss_filter_attr we discard it from the user theme.json.

## How has this been tested?
Try to inject unsafe rules, e.g: SVG backgrounds in the gradients, and verify these rules are removed.
To make it easier to test, this PR is escaping styles submitted by an admin "! current_user_can( 'unfiltered_html' ) || true " the  "|| true " should be removed before the merge.

### CODE REVIEW

File: lib/global-styles.php
Comment (Soean) [inline lib/global-styles.php:323]: This if condition is always true.

Comment (jorgefilipecosta) [reply lib/global-styles.php:323]: Hi @Soean, that was done on purpose to make it easier to test the PR, without needing a multisite where admin doesn't have an unfiltered HTML capability, or without using a plugin that removes the capability. Before merging the PR the "|| true" should be removed.

**Comment by nosolosw** [comment]
Hey, I've taken a first look at this. I want to review further the part of this PR that deals with hooking into the `content_save_pre` filter, so will give feedback about this a bit later.

Here's my feedback about the other two things this PR does:

- I like that the implementation of `remove_insecure_properties` considers more use cases than what we have in `process_key` with the `should_escape` flag. It doesn't seem that we need both mechanisms, though. I'm also thinking that dealing with this via constructor is a better way than having a separate public API method. So, what would you think of creating a separate PR that just updates the code in `process_key`?

- For the `safecss_filter_attr_allow_css` filter. I understand this is trying to fix the situation in which a block style has a custom value for presets. At the moment, this only happens for the font-family in the site-editor. We also have a highly related issue with link colors not working for non-admin roles https://github.com/WordPress/gutenberg/issues/25151. In the interest of velocity and having nice, well-scoped PR, do you think we can take this code, revive/combine it with what we had at https://github.com/WordPress/gutenberg/pull/25411 and create a separate PR to deal with this?

File: lib/global-styles.php
Comment (nosolosw) [reply lib/global-styles.php:323]: Perhaps we can remove the `true` condition from this code and add it to the testing instructions? We need to test both that users can and can't store that data depending on their capabilities.

**Comment by jorgefilipecosta** [comment]
> For the safecss_filter_attr_allow_css filter. I understand this is trying to fix the situation in which a block style has a custom value for presets. At the moment, this only happens for the font-family in the site-editor



Hi @nosolosw, it happens on every preset. When we apply a background color for global the style generated is:
```
:root {
	background-color: var(--wp--...)
}
```

You are right it is a different issue, would you prefer to separate that part of this PR in a dependent PR? We would need to merge that PR first.

**Comment by jorgefilipecosta** [comment]
> I like that the implementation of remove_insecure_properties considers more use cases than what we have in process_key with the should_escape flag. It doesn't seem that we need both mechanisms, though. I'm also thinking that dealing with this via constructor is a better way than having a separate public API method. So, what would you think of creating a separate PR that just updates the code in process_key?

Hi @nosolosw, I guess we could remove the process_key function and the process_key calls? 
ll our escaping needs seem to be covered in this PR. And the current process_key methodology does not allow unfiltered styles by users with that ability.
The approach used in this PR can not be applied inside a process_key call. In this PR I don't escape individual style keys, I just generate the styles using the normal function compute_style_properties, and then allow the keys that generated a valid safe style, cleaning all the rest (keys that did not output any style or unsafe keys).

**Review by oandregal — APPROVED** [review]
Was able to read the other part of the code and this seems fine. I wasn't able to test all the use cases this addresses, though (don't know how to test the import case, for example). I think this is important to merge and test widely in the coming days, so I'm approving.

I'd like to see these follow-ups (the smaller the PR the easier/quicker to review/land):

- Add tests.

- The presets listed [here](https://github.com/WordPress/gutenberg/pull/28061/files#diff-0b8e98dead5542e95159bd3c482158111d3f16637e7eef9af13373358fa42dedR361) need to be taken from the schema, so adding a new preset doesn't require touching multiple files (so people will forget).

- We need to merge process_key & remove_insecure_properties.

I've also got a couple of questions about how we use `isGlobalStylesUserThemeJSON`:

- We seem to use this flag to signal both that the data is a CPT coming from the user & that the data has been sanitized. Do you think it makes sense to have separate flags instead? `isUserData` and `isSanitized` for example.

- On the other hand, I haven't seen that we check that a post is sanitized before using it, so I'm not sure we need [this](https://github.com/WordPress/gutenberg/pull/28061/files#diff-41112e4d1e1b9d7fe224c7c598fb56d7ed165d57335008129e1c4658252fb037R238) and [this](https://github.com/WordPress/gutenberg/pull/28061/files#diff-47f92ee7ec8f0962669d2dd06031dbbae3e37244cb96f77c99160ca756104c85R139). What do you think?

**Comment by nosolosw** [comment]
I've started to create follow-ups: https://github.com/WordPress/gutenberg/pull/28171

**Comment by jorgefilipecosta** [comment]
Hi @nosolosw,

> I've also got a couple of questions about how we use isGlobalStylesUserThemeJSON:

This flag is really important and we should decide well on its name as we can not easily change it. It is store in the database and we need to be back-compatible with it.

In fact, I wonder if as part of theme.json versioning it also makes sense to save a version on the database, so we are back-compatible with old user data?

> We seem to use this flag to signal both that the data is a CPT coming from the user & that the data has been sanitized. Do you think it makes sense to have separate flags instead? isUserData and isSanitized for example.


This flag indicates that is a user theme.json, if is a user theme.json it is sanitized or not depending on the user and conditions/plugins being present on the save moment. I don't think a isSanitized flag is useful at all.



> On the other hand, I haven't seen that we check that a post is sanitized before using it, so I'm not sure we need this and this. What do you think?

We need to check if the flag is preset or not. On HTML content there is no need to check because the filters always run (depending just on user capabilities). We can not run our filters unconditionally as otherwise, we may change JSON data from another plugin for example. If the flag is not preset the content can not be presumed to be safe, so before using we need to verify if the flag is there.
Otherwise, an attacker could create a JSON structure with unallowed styles but without the flag. The content would not be sanitized but would end up being used anyway and our theme.json system would output styles that the user was no permission to do it.

**Comment by nosolosw** [comment]
Follow-up https://github.com/WordPress/gutenberg/pull/28188

@jorgefilipecosta would you think you are able to add tests for this in a new PR? I can take care of the other follow-ups in some PRs I'm working on.

**Comment by jorgefilipecosta** [comment]
Hi @nosolosw,
Thank you for taking care of the other follow-ups.
I will work on adding the test cases.

**Pull request merged**
